### PR TITLE
Sils/prerename

### PR DESCRIPTION
**Pull request by sils1297**

### CODE REVIEW

File: docs/Users/Install.rst [hunk 1]
```diff
@@ -166,6 +166,28 @@ https://docs.python.org/3.3/install/#alternate-installation.
     http://stackoverflow.com/questions/14149422/using-pip-behind-a-proxy
     for further clarification.
 
+Error handling
+~~~~~~~~~~~~~~~~
```
Comment (Adrianzatreanu) [inline docs/Users/Install.rst]: hm never used the usage of `~` before, what does it do?

File: docs/Users/Install.rst [hunk 2]
```diff
@@ -166,6 +166,28 @@ https://docs.python.org/3.3/install/#alternate-installation.
...
+    ::
+
+        $ pip3 install virtualenv
+        $ virtualenv -p python3 ~/venvs/coala
+        $ . ~/venv/coala/bin/activate
+        $ pip install -U pip
+        $ pip install coala-bears
+
+You have to activate this virtualenv on every terminal session you want to use
+coala though (tip: add it to bashrc!)
```
Comment (Adrianzatreanu) [inline docs/Users/Install.rst]: period after though? else this sentence has no ending.

**Comment by Adrianzatreanu** [comment]
ack 94766dd2

File: docs/Users/Install.rst (hunk 1 above)
Comment (sils1297) [inline docs/Users/Install.rst]: the same, this commit is already acked and I rebased on it because it's just not merged yet

**Comment by Adrianzatreanu** [comment]
ack 71dca8aa

**Comment by Adrianzatreanu** [comment]
ack d9586c039

File: docs/Users/Install.rst (hunk 1 above)
Comment (SanketDG) [inline docs/Users/Install.rst]: could this have a better name like `Possible Errors`?

Comment (sils1297) [inline docs/Users/Install.rst]: this commit has already been merged, file an issue

**Comment by sils1297** [comment]
@Udayan12167 you probably wanna look at this one first, these changes are almost ready to merge, I guess some test is missing somewhere. They introduce the renaming/deletion but it isn't dealt with by the actions properly which happens in the other PR.

**Comment by sils1297** [comment]
@Udayan12167 any progress on the renaming?

**Comment by Udayan12167** [comment]
Looking at this now

File: coalib/results/Diff.py [hunk 3]
```diff
@@ -115,6 +118,21 @@ def __len__(self):
...
+
+    @rename.setter
+    @enforce_signature
+    def rename(self, rename: (str, False)):
+        """
+        :param rename: False or string containing new name of file.
+        """
+        self._rename = rename
+
+    @property
```
Comment (Udayan12167) [inline coalib/results/Diff.py]: Unrelated change to this commit

**Comment by Udayan12167** [comment]
ack a31cf01

**Comment by Udayan12167** [comment]
ack 1732980

**Comment by Udayan12167** [comment]
ack 3c860d3

File: coalib/results/Diff.py [hunk 4]
```diff
@@ -176,7 +197,7 @@ def unified_diff(self):
         """
         return ''.join(difflib.unified_diff(
             self.original,
-            self.modified,
+            self.modified if not self.delete else [],
```
Comment (sils1297) [inline coalib/results/Diff.py]: `if not self.delete else []` isn't needed since `modified` already takes care of this, right?

File: tests/results/DiffTest.py [hunk 5]
```diff
@@ -93,6 +93,12 @@ def test_modified(self):
         self.uut.add_lines(0, ["0.1", "0.2"])
         self.uut.add_lines(1, ["1.1"])
         self.uut.change_line(3, "3", "3.changed")
+
+        # If delete is True then modified returns an empty list
+        self.uut.delete = True
+        self.assertEqual(self.uut.modified, [])
+        self.uut.delete = False
```
Comment (sils1297) [inline tests/results/DiffTest.py]: can you move this into `test_delete`? Don't blow up unrelated test cases so much.

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2132/commits/6cb4d9c287ee136b99932ac07dba18c73fddfb51

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2132/commits/e8e891647339cf51dbf01ed5568999a28b0dfc14

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2132/commits/ebd440892bc3f35a21fad3c44842a6b454cf0289 https://github.com/coala-analyzer/coala/pull/2132/commits/461b5e8e0816481d581844ac82620f56a00c6528 https://github.com/coala-analyzer/coala/pull/2132/commits/7d78f781a09b95d86a81a13040125fb15cff30b5 https://github.com/coala-analyzer/coala/pull/2132/commits/2d764d10911d0bbaa126c9249ba11252b669bb95 https://github.com/coala-analyzer/coala/pull/2132/commits/6dbb7b8fcbdfa849b12746da0f0104e2f82f6bc7 https://github.com/coala-analyzer/coala/pull/2132/commits/351b367821522898b2e07f2cd46006e07e941ada

**Comment by sils1297** [comment]
@Udayan12167 minimal work only required, can we get this going again?

**Comment by sils1297** [comment]
@Udayan12167 reack?

**Comment by Udayan12167** [comment]
ack a658b84 495c2f6 97b4eb8 5caf7d8 61191f6 fdefc93

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2132/commits/0d2541da4490738a945675594c8d2e25a6dff74d

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2132/commits/eb2172ed2274ad3bb635952382b9ec2fc8753e86

File: coalib/results/result_actions/ApplyPatchAction.py [hunk 6]
```diff
@@ -35,19 +36,28 @@ def apply(self,
...
-                shutil.copy2(filename, filename + ".orig")
+            if not no_orig and not isfile(pre_patch_filename + ".orig"):
+                shutil.copy2(pre_patch_filename, pre_patch_filename + ".orig")
 
-            # Write new contents
-            with open(filename, mode='w', encoding='utf-8') as file:
-                file.writelines(new_file)
+            diff = file_diff_dict[filename]
+            if diff.delete:
+                remove(pre_patch_filename)
```
Comment (sils1297) [inline coalib/results/result_actions/ApplyPatchAction.py]: what if the file was already deleted by a previous patch?

File: coalib/results/result_actions/ApplyPatchAction.py [hunk 7]
```diff
@@ -35,19 +36,28 @@ def apply(self,
...
+            else:
+                new_filename = (diff.rename
+                                if diff.rename is not False
+                                else filename)
```
Comment (sils1297) [inline coalib/results/result_actions/ApplyPatchAction.py]: shouldn't it be pre_patch_filename?

Comment (sils1297) [inline coalib/results/result_actions/ApplyPatchAction.py]: ah no this is fine

File: tests/results/result_actions/ApplyPatchActionTest.py [hunk 8]
```diff
@@ -91,6 +91,51 @@ def test_apply_orig_option(self):
 
             self.assertEqual(file_dict, expected_file_dict)
 
+    def test_apply_rename(self):
+        uut = ApplyPatchAction()
+        with make_temp() as f_a:
+            file_dict = {
+                f_a: ["1\n", "2\n", "3\n"]
+                }
```
Comment (sils1297) [inline tests/results/result_actions/ApplyPatchActionTest.py]: those three lines can be put in one

Comment (sils1297) [inline tests/results/result_actions/ApplyPatchActionTest.py]: makes it less scary

File: tests/results/result_actions/ApplyPatchActionTest.py [hunk 9]
```diff
@@ -91,6 +91,51 @@ def test_apply_orig_option(self):
...
+            file_diff_dict = {}
+            diff = Diff(file_dict[f_a], delete=True)
+            uut.apply(Result("origin", "msg", diffs={f_a: diff}),
+                      file_dict,
+                      file_diff_dict)
+
+            self.assertFalse(isfile(f_a))
+            open(f_a, 'w')
```
Comment (sils1297) [inline tests/results/result_actions/ApplyPatchActionTest.py]: try applying another patch that has the delete flag also set, there should be no conflict and nothing happening because all is well (but I think your current routine would attemt to delete it again)

File: coalib/results/result_actions/ApplyPatchAction.py [hunk 10]
```diff
@@ -35,19 +36,30 @@ def apply(self,
...
+            diff = file_diff_dict[filename]
+            if isfile(pre_patch_filename):
+                if diff.delete:
+                    remove(pre_patch_filename)
+                else:
+                    new_filename = (diff.rename
+                                    if diff.rename is not False
+                                    else filename)
+                    with open(
+                            new_filename, mode='w', encoding='utf-8') as file:
```
Comment (sils1297) [inline coalib/results/result_actions/ApplyPatchAction.py]: this is a bit misformatted

Comment (sils1297) [inline coalib/results/result_actions/ApplyPatchAction.py]: rather break after the comma if it really doesn't fit

**Comment by sils1297** [comment]
@Udayan12167 your tests failed

**Comment by Udayan12167** [comment]
It's just that one test failing for a some unknown reason

**Comment by sils1297** [comment]
@Udayan12167 if you need help investigating this, please ask people about it

**Comment by Udayan12167** [comment]
reack 3299b99 596f9a5 d931bc7 c979e0c 14a7038 6cdf7b4 ef3a824 326c1d7 73a2faa

File: tests/results/result_actions/ApplyPatchActionTest.py [hunk 11]
```diff
@@ -91,6 +92,53 @@ def test_apply_orig_option(self):
...
+
+            diff = Diff(file_dict[f_a])
+            diff.change_line(2, "2\n", "2_changed\n")
+            uut.apply(Result("origin", "msg", diffs={f_a: diff}),
+                      file_dict,
+                      file_diff_dict)
+            self.assertTrue(isfile(f_a+".renamed.orig"))
+
+            for filename in file_diff_dict:
+                file_dict[filename] = file_diff_dict[filename].modified
```
Comment (sils1297) [inline tests/results/result_actions/ApplyPatchActionTest.py]: could you check if the contents of the actual file are good? Also shouldn't the filename in the expected dict be `f_a + '.renamed'`?

File: tests/results/result_actions/ApplyPatchActionTest.py [hunk 12]
```diff
@@ -91,6 +92,53 @@ def test_apply_orig_option(self):
...
+            self.assertTrue(isfile(f_a+".orig"))
+            os.remove(f_a+".orig")
+
+            diff = Diff(file_dict[f_a])
+            diff.change_line(3, "3\n", "3_changed\n")
+            uut.apply(Result("origin", "msg", diffs={f_a: diff}),
+                      file_dict,
+                      file_diff_dict)
+            self.assertFalse(isfile(f_a+".orig"))
+            open(f_a, 'w').close()
```
Comment (sils1297) [inline tests/results/result_actions/ApplyPatchActionTest.py]: maybe add a comment why this is needed

**Comment by sils1297** [comment]
and rebase

**Comment by Udayan12167** [comment]
reack a0a42a0 413b241 1cf84fc d1981a6 c47ae3f 4709ef1 a52ec68 db19c99 45533d2

File: tests/results/result_actions/ApplyPatchActionTest.py [hunk 13]
```diff
@@ -91,6 +92,58 @@ def test_apply_orig_option(self):
...
+
+            diff = Diff(file_dict[f_a])
+            diff.change_line(2, "2\n", "2_changed\n")
+            uut.apply(Result("origin", "msg", diffs={f_a: diff}),
+                      file_dict,
+                      file_diff_dict)
+            self.assertTrue(isfile(f_a+".renamed.orig"))
+
+            file_dict = {f_a+".renamed": open(f_a+".renamed").readlines()}
+            print(file_dict)
```
Comment (sils1297) [inline tests/results/result_actions/ApplyPatchActionTest.py]: remove debug print please

Comment (sils1297) [inline tests/results/result_actions/ApplyPatchActionTest.py]: otherwise it's good, just reack it after removing the print and merge it?

**Comment by Udayan12167** [comment]
reack 110e783

**Comment by Udayan12167** [comment]
@rultor merge
